common/src/main/java/org/conscrypt/ConscryptFileDescriptorSocket.java

```diff
@@ -216,6 +216,9 @@ public void startHandshake() throws IOException {
 
             try {
                 ssl.doHandshake(Platform.getFileDescriptor(socket), getSoTimeout());
+
+                // Update the session from the current state of the SSL object.
+                sslSession.onHandshakeCompleted(getHostnameOrIP(), getPort());
```

Comment:
You are not guaranteed the handshake is *fully* completed, but that is the wrong signal to be listening for here. The immediate motivation was the old place had no way to route the exception out.

Where it was originally was also incorrect in the case of False Start (certs got filled in too late) but just didn't matter because in those cases the callback got run so the certificates didn't need to be filled in. As in the commit message, when `SSL_do_handshake` completes successfully, properties are available and most code should treat that as handshake "complete".

I say "complete" in quotes because False Start and later 0-RTT rely on the handshake "completing" early w.r.t. properties being available and I/O working. Full completion will happen later, the exact semantics of which are specific to the feature being enabled that relies on it (False Start may mostly be ignored, 0-RTT has some complicated caller obligations).

Probably `onHandshakeCompleted` should be named `onCertificatesAvailable` or something. Complete is too vague of a term. I'll add a commit to rename that function.